File: src/main/scala/demo/customer/CustomerService.scala
package demo.customer

import demo.customer.ProductType.ProductType

import scala.util.{Failure, Success, Try}

object ProductType extends Enumeration {
  type ProductType = Value
  val Book, Bluray = Value
}

class Store(
    private val inventory: Map[ProductType, Int] = Map.empty[ProductType, Int]
) {
  def removeInventory(product: ProductType, quantity: Int): Store =
    updateStore(product, inventory(product) - quantity)

  def hasEnoughInventory(product: ProductType, quantity: Int): Boolean =
    getInventoryFor(product) >= quantity

  def getInventoryFor(product: ProductType): Int =
    inventory.getOrElse(product, 0)

  def addInventory(product: ProductType, quantity: Int): Store =
    updateStore(product, inventory.getOrElse(product, 0) + quantity)

  private def updateStore(product: ProductType, newQuantity: Int): Store =
    new Store(inventory + (product -> newQuantity))
}

object CustomerService {
  def purchase(
      store: Store,
      product: ProductType,
      quantity: Int
  ): Try[Store] =
    if (!store.hasEnoughInventory(product, quantity))
      Failure(new IllegalArgumentException("Not enough inventory"))
    else Success(store.removeInventory(product, quantity))
}

File: java/src/main/java/customer/CustomerService.java
package customer;

import io.vavr.collection.Map;
import io.vavr.control.Try;
import lombok.AllArgsConstructor;

import static io.vavr.API.Failure;
import static io.vavr.API.Success;

enum ProductType {
    Bluray,
    Book
}

@AllArgsConstructor
class Store {
    private final Map<ProductType, Integer> inventory;

    public Store removeInventory(ProductType product, Integer quantity) {
        return updateStore(product, getInventoryFor(product) - quantity);
    }

    public Store addInventory(ProductType product, Integer quantity) {
        return updateStore(product, getInventoryFor(product) + quantity);
    }

    public boolean hasEnoughInventory(ProductType product, Integer quantity) {
        return getInventoryFor(product) >= quantity;
    }

    private Integer getInventoryFor(ProductType product) {
        return inventory.getOrElse(product, 0);
    }

    private Store updateStore(ProductType product, Integer newQuantity) {
        return new Store(inventory.put(product, newQuantity));
    }
}

public class CustomerService {
    public Try<Store> purchase(
            Store store,
            ProductType product,
            Integer quantity
    ) {
        return (!store.hasEnoughInventory(product, quantity)) ?
                Failure(new IllegalArgumentException("Not enough inventory")) :
                Success(store.removeInventory(product, quantity));
    }
}
